util/rate_limiter.cc

```diff
@@ -136,96 +136,85 @@ void GenericRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
   Req r(bytes, &request_mutex_);
   queue_[pri].push_back(&r);
 
+  // A thread representing a queued request coordinates with other such threads.
+  // There are two main duties.
+  //
+  // (1) Waiting for the next refill time.
+  // (2) Refilling the bytes and granting requests.
+  //
+  // To simplify the implementation, the duties can be performed by any thread.
```

Comment:
Just to understand the "simplification" effects in "coding", does the "simplification" result in the following major coding changes:
(1) Remove the non-leader thread else-branch on [L218](https://github.com/facebook/rocksdb/pull/8602/files#diff-64f39d34591bb466f92056637ccb1b88009806eabf366a59eca5b92d27fea2c8L218) 
(2) Cut down the "Spontaneous wake up" else-branch on [L213](https://github.com/facebook/rocksdb/pull/8602/files#diff-64f39d34591bb466f92056637ccb1b88009806eabf366a59eca5b92d27fea2c8L213)
(3) Facilitate the destructor clean up (i.e, if (stop_) {...} outside of the while loop on [L139](https://github.com/facebook/rocksdb/pull/8602/files#diff-64f39d34591bb466f92056637ccb1b88009806eabf366a59eca5b92d27fea2c8L139)

Did I miss any other major coding changes (i.e, non-assertion change) the simplification results in?

Reply:
I can change the claim to say "reduce conceptual complexity" to make it easier to reason about edge cases, change/extend, etc. There isn't any particular line in the old code that's exciting to remove; it's more about how long it takes to convince myself it works in all cases. In the old code it took me a really long time to draw a timeline to understand how it works for just one edge case yet there were so many.

Reply:
I agree that conceptual complexity in reasoning about the timeline of different threads and edge cases is the pain point of the previous impl, which is also why we need this PR.

Suggestion: let's add the "why" we need to simplify the previous algorithm in the PR summary too.

Reply:
> imo the main simplification to the logic is there is no choice of leader anymore. 
I think the key reason why we can do this is - the constraint of a leader (i.e, the heavy-lifting thread) being the front request in the queue, which is the only leader constraint being relaxed in this PR, is unnecessary for correctness. The reason is that the correctness in our Refill() means - if there is any ungranted request, we make sure there exists one and only one thread to refill the bytes and grant requests when time is right. None of these requires the front-queued constraint.

In addition to that simplification, I will also argue the current coding style of separating the code paths needed to be read by readers in order to trace different threads is also an improvement to simplicity and readability.
 
Suggestion: 
Let's add a PR summary of "what" the simplification are and "why" we can simplify in this way without violate the correctness. 
We can also mention losing the potential benefit of saving one wake-up through leader being the front request, as you have described, is acceptable.

Reply:
Suggestion:
How about we move the paragraph from L148 to 183 out of the code comment section and structure them into the PR summary? 

(1) For example, L148-L152 addresses "why" we can simplify the existing algorithm, echoing back to my previous suggestion
> ... and "why" we can simplify in this way without violating the correctness?

(2) L153-163 addresses some code design details on how we reduce unnecessary wake-ups since we know from L149-150 that there could potentially be many wasted wake-ups due to only one of them will hold the mutex and proceed.

(3) L167-170 addresses some code design details on the "front-queued" property and its benefit. Moving these lines to PR summary also helps describe a change in this PR (where "front-queued" property typically holds but not always) from existing algorithm (where the "front-queued" property always holds). 

It also naturally leads to the argument you have in PR that the change is acceptable based on your performance result.

(4) I am a bit confused on some part from L172-183 - see my question along with the suggestion for these lines below.

The reason why these lines can be moved out of the code comment is that, IMO, L142-146 and your re-written version of the loop starting from L184 is self-explanatory. Even for a cautious reader tempting to hand-trace different threads to prove the correctness of the program by him/herself, the current version of impl doesn't seem to be too hard for them to do so. Even it ends up being hard, what they probably need is an illustration of possible timelines consisting of different threads based on different cases in the loop, instead of L148 to 183.